main/src/ca/uwaterloo/flix/language/ast/SourceLocation.scala

```diff
@@ -40,18 +42,19 @@ object SourceLocation {
   *
   * Specifically, we:
   *
+  * - Use a nullable reference to `ParserInput`.
   * - Use a `Boolean` to represent whether a source location is real (true) or synthetic (false).
   * - Use `Short`s instead of `Int`s to represent column offsets (i.e. `beginCol` and `endCol`).
   *
-  * @param input        the parser input.
+  * @param input        the nullable parser input.
   * @param source       the source input.
   * @param isReal       true if real location, false if synthetic location.
   * @param beginLine    the line number where the entity begins.
   * @param beginCol     the column number where the entity begins.
   * @param endLine      the line number where the entity ends.
   * @param endCol       the column number where the entity ends.
   */
-case class SourceLocation(input: Option[ParserInput], source: Source, isReal: Boolean, beginLine: Int, beginCol: Short, endLine: Int, endCol: Short) {
+case class SourceLocation(@Nullable input: ParserInput, source: Source, isReal: Boolean, beginLine: Int, beginCol: Short, endLine: Int, endCol: Short) {
```

Comment:
Can we make `input` private so the null ability cannot leak?

Reply:
Case classes cannot have private fields.

Reply:
Ah they can. But I can only make it private for source location.